src/test/transaction_tests.cpp

```diff
@@ -266,8 +271,8 @@ BOOST_AUTO_TEST_CASE(tx_valid)
             }
 
             // Check that flags are maximal: transaction should fail if any unset flags are set.
-            for (auto flags_excluding_one : ExcludeIndividualFlags(verify_flags)) {
-                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~flags_excluding_one, txdata, strTest, /* expect_valid */ false)) {
+            for (auto submask_flags : GenerateAllSubMasks(verify_flags)) {
+                if (!CheckTxScripts(tx, mapprevOutScriptPubKeys, mapprevOutValues, ~submask_flags, txdata, strTest, /* expect_valid */ false)) {
                     BOOST_ERROR("Too many flags unset: " << strTest);
                 }
             }
```

Comment:
If I understand the intent of this code correctly, the idea is that given a set of chosen flags, we want to check that every possible superset of flags doesn't succeed. If that's so, I believe there's a problem with the current `FlipAll( GenerateAllSubMasks( flags ) )` logic for computing all possible supersets.

Eg, assuming 3-bit words, for `flags = 100`, we'll generate all `flags` sub masks `{000}`, and then flip them `{111}`. However, what we really want is to generate all the `~flags = 011` sub masks `{010, 001, 000}`, and then flip them `{101, 110, 111}`.

In short, there's a missing bit flip before computing all the sub masks. Ie. it should be
```
FlipAll( GenerateAllSubMasks( ~flags ) )
```

With that said, there's still the issue that flipping flag bits doesn't interact well with the `TrimFlags` operation done inside `GenerateAllSubMasks`.

Eg. if the only flag bits were`VERIFY_P2SH` and `VERIFY_WITNESS` (in that order), for `flags = 00` we'd generate all the `~flags = 11` sub masks `{10, 01, 00}`, trim them `{10, 00, 00}` (`01` is mapped to `00`), remove duplicates `{10, 00}`, and finally flip them `{01, 11}`. However, we ended up with an invalid combination (`01`) when the correct result would have been `{10, 11}`.

More generally, after fixing the initial missing flip, the code would look something like
```
FlipAll( Unique( TrimFlags( GenerateAllSubMasksWithoutTrimming( ~flags ) ) ) )
```
when what we really want is something like:
```
for flag_superset in FlipAll( GenerateAllSubMasksWithoutTrimming( ~flags ) ) ) ):
    if flag_superset == TrimFlags( flag_superset ):
        # use flag_superset
```

Reply:
yeah you're totally right; i've been playing with it a bit as well and noticed that it seems to not be quite right (thanks @jonatack for pointing it out).

The logic is definitely off, and I think it may be intractible in this case since GenerateAllSubMasksWithoutTrimming(~flags) ends up being pretty big... with 20 it might be ok, but eventually there would be too many flags.

Reply:
Makes sense. Maybe bound the amount of flipped flags? That way you can test some flag interactions, but bound the exponential growth.